### PR TITLE
Entities: better document Marin entity

### PR DESCRIPTION
**Pull request by kemenaran**
Start documenting Marin's entity – especially how this entity renders Link turning in its bed at the game opening.

### CODE REVIEW

File: src/code/entities/05__helpers_1.asm [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    pop  hl                                       ;; 05:54FF $E1
+    and  1                                        ;; 05:5500 $E6 $01
+    or   [hl]                                     ;; 05:5502 $B6
+
+    jp   SetEntitySpriteVariant                   ;; 05:5503 $C3 $0C $3B
+
+ShouldLinkTalkToEntity_05::
+    ;
+    ; Check if Link is vertically close from the entity
```
Comment (tobiasvl) [inline src/code/entities/05__helpers_1.asm:73]:
```suggestion
    ; Check if Link is vertically close to the entity
```

File: src/code/entities/05__helpers_1.asm [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    ld   hl, hActiveEntityPosY                    ;; 05:551B $21 $EF $FF
+    sub  [hl]                                     ;; 05:551E $96
+    add  $14                                      ;; 05:551F $C6 $14
+    cp   $28                                      ;; 05:5521 $FE $28
+.checkYPosEnd
+
+    jr   nc, .dontTalk                            ;; 05:5523 $30 $44
+
+    ;
+    ; Check if Link is horizontally close from the entity
```
Comment (tobiasvl) [inline src/code/entities/05__helpers_1.asm:100]:
```suggestion
    ; Check if Link is horizontally close to the entity
```

File: src/code/entities/05_marin.asm [hunk 3]
```diff
@@ -85,25 +85,25 @@ MarinEntityHandler::
     and  a                                        ;; 05:4E74 $A7
     jp   nz, MarinEntityHandler_Indoor            ;; 05:4E75 $C2 $CE $51
 
-    ; Didn't found sword yet: remove the entity
+    ; Didn't found sword yet: remove the outdoor entity
```
Comment (tobiasvl) [inline src/code/entities/05_marin.asm:88]:
```suggestion
    ; Didn't find sword yet: remove the outdoor entity
```

File: src/code/entities/05_marin.asm [hunk 4]
```diff
@@ -134,21 +134,25 @@ MarinEntityHandler::
...
+    call GetEntityDirectionToLink_05              ;; 05:4EC9 $CD $24 $7B
     ld   hl, wEntitiesDirectionTable              ;; 05:4ECC $21 $80 $C3
     add  hl, bc                                   ;; 05:4ECF $09
     ld   [hl], e                                  ;; 05:4ED0 $73
-
 .jr_4ED1
-    call func_005_54EA                            ;; 05:4ED1 $CD $EA $54
+
+    ;
+    ; Spaw music notes when Marin is singing
```
Comment (tobiasvl) [inline src/code/entities/05_marin.asm:144]:
```suggestion
    ; Spawn music notes when Marin is singing
```

File: src/code/entities/05_marin.asm [hunk 5]
```diff
@@ -644,177 +656,215 @@ func_005_51BC::
...
     call ShouldLinkTalkToEntity_05                ;; 05:51F2 $CD $06 $55
     ret  nc                                       ;; 05:51F5 $D0
+    jp_open_dialog Dialog1D7 ; "At the beach..."  ;; 05:51F6 $3E $D7
+.letterEnd
 
-    jp_open_dialog Dialog1D7                      ;; 05:51F6 $3E $D7
-
-.jr_51FB
+    ;
+    ; After Link retrieved its sword, Marin is no longer home. Remove the entity.
```
Comment (tobiasvl) [inline src/code/entities/05_marin.asm:701]:
```suggestion
    ; After Link retrieves his sword, Marin is no longer home. Remove the entity.
```

File: src/code/entities/05_marin.asm [hunk 6]
```diff
@@ -644,177 +656,215 @@ func_005_51BC::
...
+    ;
     ld   a, [wSwordLevel]                         ;; 05:51FB $FA $4E $DB
     and  a                                        ;; 05:51FE $A7
     jp   nz, ClearEntityStatus_05                 ;; 05:51FF $C2 $4B $7B
 
+    ; After Link retrieved its shield, set the state directly to MARIN_STATE_3
```
Comment (tobiasvl) [inline src/code/entities/05_marin.asm:707]:
```suggestion
    ; After Link retrieves his shield, set the state directly to MARIN_STATE_3
```

**Review by tobiasvl — APPROVED** [review]

File: src/code/entities/05_marin.asm [hunk 7]
```diff
@@ -644,177 +656,215 @@ func_005_51BC::
...
     ldh  [hActiveEntityPosX], a                   ;; 05:52B3 $E0 $EE
     ldh  [hLinkPositionX], a                      ;; 05:52B5 $E0 $98
     ld   a, $34                                   ;; 05:52B7 $3E $34
     ldh  [hActiveEntityVisualPosY], a             ;; 05:52B9 $E0 $EC
     ldh  [hLinkPositionY], a                      ;; 05:52BB $E0 $99
+
     ld   a, $02                                   ;; 05:52BD $3E $02
     ldh  [hLinkInteractiveMotionBlocked], a       ;; 05:52BF $E0 $A1
+
     ld   a, LINK_ANIMATION_STATE_NO_UPDATE        ;; 05:52C1 $3E $FF
```
Comment (daid) [inline src/code/entities/05_marin.asm:835]: While this says "no update", I think this actually hides link completely. Might need a test if this only hides link if you directly set this on room load or if you can always use this to hide the link sprite.

**Review by daid — APPROVED** [review]
Looks good, not seeing anything that doesn't match my observations.

File: src/code/entities/05_marin.asm (hunk 7 above)
Comment (kemenaran) [reply src/code/entities/05_marin.asm]: Yes, it does hide Link completely. It is also used when Link is falling down a pit, at the end of the animation, when he is no longer visible.

I renamed it to LINK_ANIMATION_STATE_HIDDEN.

**Pull request merged**